File: projects/07-Treaps/implementation/treaps.cpp
#include <iostream>
#include <vector>
#include <random>
#include <limits>
using namespace std;

/*
    Treap = Tree + Heap

    A treap is a randomized binary search tree.

    It maintains TWO properties at the same time:

    1. BST property by key:
       - all keys in the left subtree are smaller
       - all keys in the right subtree are larger

    2. Heap property by priority (max-heap here):
       - each node's priority is >= its children's priorities

    The key determines where a node belongs in sorted order.
    The priority determines how high the node rises in the tree.

    Random priorities make the tree balanced in expectation,
    so insert/search/erase are expected O(log n).
*/

class Treap {
private:
    /*
        Each node stores:
        - key: the BST ordering value
        - priority: the heap ordering value
        - left/right child pointers
    */
    struct Node {
        int key;
        int priority;
        Node* left;
        Node* right;

        Node(int k, int p) : key(k), priority(p), left(nullptr), right(nullptr) {}
    };

    Node* root;

    /*
        Random number generator for priorities.

        We use a large integer range so equal priorities are unlikely.
    */
    mt19937 rng;
    uniform_int_distribution<int> dist;

private:
    /*
        split(node, key)

        Splits a treap into two treaps:
        - left treap contains all keys <= key
        - right treap contains all keys > key

        This function is one of the most important treap tools!

        Why it works:
        - Because the input is already a valid treap,
          we only need to recursively split one subtree at a time.
        - The BST property tells us which side keys belong to.
        - The heap property is preserved automatically by reconnecting subtrees carefully.

        Returns:
            pair<left_treap_root, right_treap_root>
    */
    pair<Node*, Node*> split(Node* node, int key) {
        if (node == nullptr) {
            return {nullptr, nullptr};
        }

        if (node->key <= key) {
            /*
                Current node belongs in the LEFT result.

                But some nodes in node->right might still be <= key,
                so we recursively split the right subtree.
            */
            auto [middle, rightPart] = split(node->right, key);
            node->right = middle;
            return {node, rightPart};
        } else {
            /*
                Current node belongs in the RIGHT result.

                But some nodes in node->left might still be <= key,
                so we recursively split the left subtree.
            */
            auto [leftPart, middle] = split(node->left, key);
            node->left = middle;
            return {leftPart, node};
        }
    }

    /*
        merge(left, right): Merges two treaps into one treap.

        Note: Every key in 'left' must be <= every key in 'right'.

        If that is true, then we can merge based on priority:
        - whichever root has higher priority becomes the new root
        - recursively merge the remaining subtree

        Why it works:
        - BST property holds because all left keys stay before all right keys
        - heap property holds because higher priority root stays on top
    */
    Node* merge(Node* left, Node* right) {
        if (left == nullptr) return right;
        if (right == nullptr) return left;

        if (left->priority > right->priority) {
            /*
                Left root has higher priority, so it becomes the new root.
                Its right child should be the merge of:
                - left's original right subtree
                - the entire right treap
            */
            left->right = merge(left->right, right);
            return left;
        } else {
            /*
                Right root has higher priority, so it becomes the new root.
                Its left child should be the merge of:
                - the entire left treap
                - right's original left subtree
            */
            right->left = merge(left, right->left);
            return right;
        }
    }

    /*
        search(node, key)

        Standard BST lookup:
        - if key is smaller, go left
        - if key is larger, go right
        - if equal, found it

        Priority does NOT matter for searching.
        Priority only affects shape, not lookup direction.
    */
    bool search(Node* node, int key) const {
        if (node == nullptr) return false;

        if (key == node->key) return true;
        if (key < node->key) return search(node->left, key);
        return search(node->right, key);
    }

    /*
        erase(node, key)

        Deletes one key from the treap.

        We first search like in a BST.
        Once we find the node:
        - remove it
        - replace it with merge(left_subtree, right_subtree)

        This works because:
        - every key in left subtree < node->key
        - every key in right subtree > node->key
        So merging those two subtrees preserves BST order.
    */
    Node* erase(Node* node, int key) {
        if (node == nullptr) return nullptr;

        if (key < node->key) {
            node->left = erase(node->left, key);
        } else if (key > node->key) {
            node->right = erase(node->right, key);
        } else {
            /*
                Found the node to delete.

                Merge its two children to reconnect the tree.
            */
            Node* merged = merge(node->left, node->right);
            delete node;
            return merged;
        }

        return node;
    }

    /*
        unite(a, b): Combines two arbitrary treaps into one.

        Correct idea:
        - choose the root with higher priority
        - split the other treap by that root's key
        - recursively unite matching left parts and right parts

        This preserves:
        - BST property
        - heap property
    */
    Node* unite(Node* a, Node* b) {
        if (a == nullptr) return b;
        if (b == nullptr) return a;

        /*
            Ensure 'a' has the higher-priority root.
        */
        if (a->priority < b->priority) {
            swap(a, b);
        }

        /*
            Split b around a->key:
            - leftB contains keys <= a->key
            - rightB contains keys > a->key

            Since this implementation uses UNIQUE keys only,
            we want:
            - strictly smaller keys on the left
            - strictly larger keys on the right

            Because split() puts <= key on the left,
            duplicates (if any) would land in leftB.
            But our public insert() prevents duplicates,
            so this is okay.
        */
        auto [leftB, rightB] = split(b, a->key);

        a->left = unite(a->left, leftB);
        a->right = unite(a->right, rightB);

        return a;
    }

    /*
        inorder(node, result)

        Standard inorder traversal:
        left -> current -> right

        In a BST, inorder traversal gives keys in sorted order.
    */
    void inorder(Node* node, vector<int>& result) const {
        if (node == nullptr) return;
        inorder(node->left, result);
        result.push_back(node->key);
        inorder(node->right, result);
    }

    /*
        validate(node, low, high)

        Checks both treap invariants:

        1. BST property:
           low < node->key < high

        2. Heap property:
           child's priority must not exceed parent's priority

        This is useful for debugging and testing.
    */
    bool validate(Node* node, long long low, long long high) const {
        if (node == nullptr) return true;

        if (!(low < node->key && node->key < high)) {
            return false;
        }

        if (node->left != nullptr && node->left->priority > node->priority) {
            return false;
        }
        if (node->right != nullptr && node->right->priority > node->priority) {
            return false;
        }

        return validate(node->left, low, node->key) &&
               validate(node->right, node->key, high);
    }

    /*
        clear(node)

        Recursively deletes all nodes to avoid memory leaks.
    */
    void clear(Node* node) {
        if (node == nullptr) return;
        clear(node->left);
        clear(node->right);
        delete node;
    }

public:
    /*
        Constructor:
        - empty treap
        - initialize RNG
    */
    Treap() : root(nullptr), rng(random_device{}()), dist(1, 1000000000) {}

    /*
        Destructor:
        frees all allocated nodes
    */
    ~Treap() {
        clear(root);
    }

    /*
        Public insert(key)

        This version enforces UNIQUE keys.
        If key already exists, do nothing.

        Algorithm:
        1. split the treap by key
        2. create new node
        3. merge left + new node
        4. merge result + right
    */
    void insert(int key) {
        if (search(key)) {
            return;  // prevent duplicates
        }

        int priority = dist(rng);
        Node* newNode = new Node(key, priority);

        auto [leftPart, rightPart] = split(root, key);
        Node* temp = merge(leftPart, newNode);
        root = merge(temp, rightPart);
    }

    /*
        Public erase(key)
    */
    void erase(int key) {
        root = erase(root, key);
    }

    /*
        Public search(key)
    */
    bool search(int key) const {
        return search(root, key);
    }

    /*
        Public unite(other)

        Combines 'other' into this treap.

        After union, we set other's root to nullptr so that
        its destructor does not delete nodes now owned by this treap.
    */
    void unite(Treap& other) {
        root = unite(root, other.root);
        other.root = nullptr;
    }

    /*
        Public inorder()

        Returns all keys in sorted order.
    */
    vector<int> inorder() const {
        vector<int> result;
        inorder(root, result);
        return result;
    }

    /*
        Public validate()

        Returns true if both BST and heap properties hold.
    */
    bool validate() const {
        return validate(root,
                        numeric_limits<long long>::lowest(),
                        numeric_limits<long long>::max());
    }
};

/*
    Helper function to print a vector nicely.
*/
void printVector(const vector<int>& v) {
    for (int x : v) {
        cout << x << " ";
    }
    cout << "\n";
}

/*
    Example usage / basic test
*/
int main() {
    int passed = 0;
    int total = 4;

    // Test 1
    Treap t1;
    t1.insert(5);
    t1.insert(3);
    t1.insert(7);
    t1.insert(2);
    t1.insert(4);

    bool test1 = t1.search(5) && t1.search(2) && !t1.search(6) && t1.validate();
    cout << "Test 1: basic insert/search "
         << (test1 ? "PASS" : "FAIL") << "\n";
    if (test1) passed++;

    // Test 2
    vector<int> expected2 = {2, 3, 4, 5, 7};
    vector<int> actual2 = t1.inorder();

    bool test2 = (actual2 == expected2);
    cout << "Test 2: inorder traversal "
         << (test2 ? "PASS" : "FAIL") << "\n";

    if (!test2) {
        cout << "Expected: ";
        printVector(expected2);
        cout << "Got:      ";
        printVector(actual2);
    }
    if (test2) passed++;

    // Test 3
    t1.erase(3);
    t1.erase(42);

    vector<int> expected3 = {2, 4, 5, 7};
    vector<int> actual3 = t1.inorder();

    bool test3 = (actual3 == expected3) && !t1.search(3) && t1.validate();
    cout << "Test 3: erase "
         << (test3 ? "PASS" : "FAIL") << "\n";

    if (!test3) {
        cout << "Expected: ";
        printVector(expected3);
        cout << "Got:      ";
        printVector(actual3);
    }
    if (test3) passed++;

    // Test 4
    Treap a;
    Treap b;

    a.insert(10);
    a.insert(20);
    a.insert(30);

    b.insert(5);
    b.insert(15);
    b.insert(25);

    a.unite(b);

    vector<int> expected4 = {5, 10, 15, 20, 25, 30};
    vector<int> actual4 = a.inorder();

    bool test4 = (actual4 == expected4) && a.validate();
    cout << "Test 4: unite "
         << (test4 ? "PASS" : "FAIL") << "\n";

    if (!test4) {
        cout << "Expected: ";
        printVector(expected4);
        cout << "Got:      ";
        printVector(actual4);
    }
    if (test4) passed++;

    cout << "\nFinal: " << passed << " / " << total << " tests passed.\n";
    return (passed == total) ? 0 : 1;
}

File: projects/07-Treaps/implementations/treaps.java
import java.util.Random;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class treaps {
    /*
    a treap is a binary search tree where each node also has a random priority.
    the tree is heap ordered by priority, while also maintaining the binary search tree property by key.
    insertions and deletions are done by splitting and merging the treap based on the priorities and keys of the nodes.
    this allows for efficient insertions, deletions, and searches while keeping the tree balanced on average.
    */

    // we need a random number generator to assign priorities to the nodes
    private static final Random rand = new Random();

    // we simply define a node to represent the nodes in the treap
    static class Node {
        int key;
        int priority;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.priority = rand.nextInt();
        }
    }

    private Node root;

    /*
    Splits the treap into two treaps by key:
    - left contains all keys <= key
    - right contains all keys > key

    If root.key <= key, root belongs in the left treap, but some nodes in root.right may still be <= key,
    so we split root.right. Otherwise root belongs in the right treap, and we split root.left.
    */
    private Node[] split(Node root, int key) {
        if (root == null) {
            return new Node[]{null, null};
        }
        // root less than or equal to the key, we split the right subtree
        if (root.key <= key) {
            Node[] res = split(root.right, key);
            // we want to set the root.right to be all the values in the right subtree that are less than or equal to the key we are at
            // this allows us to make sure none of the wrong values are in the wrong treap
            root.right = res[0];
            // we return the root as the left treap and res[1] as the right treap
            // so when we recurse back up, we will have the correct treaps with the correct priorities and keys
            return new Node[]{root, res[1]};
        } else {
            Node[] res = split(root.left, key);
            // similar logic to the above case, just simply reversed because we are splitting the left subtree instead of the right subtree

            root.left = res[1];

            return new Node[]{res[0], root};
        }
    }

    /*
    Assumes all keys in left are <= all keys in right
    the point of the merge function is to take two treaps and merge them together
    in a manner that maintains the properties of the treap,
    so we will compare the priorities of the roots of the two treaps and merge the one with the higher priority as the new root,
    and then we will merge the other treap as a child of the new root,
    we will do this recursively until we have merged the two treaps together
    */
    private Node merge(Node left, Node right) {
        if (left == null) return right;
        if (right == null) return left;

        // we call merge recursively on the child of the root with the higher priority, and then return the new root of the merged treap
        // if the left priority is higher, choose the left root 
        // we will merge the right treap as the right child of the left root, and then return the left root as the new root of the merged treap
        if (left.priority > right.priority) {
            left.right = merge(left.right, right);
            return left;
        // if the right priority is higher, choose the right root
        } else {
            right.left = merge(left, right.left);
            return right;
        }
    }

    /*
    this insert is straightforward, so if we hit the case in our recursion that the node were on's
    priority is less than the inserted node, we will split the current node by the inserted node's key and make the inserted node the new root of the subtree,
    and then we will set the left and right children of the new root to be the two treaps that we got from the split function 
    this allows us to insert our node while maintaining treap property
    */
    // we call insert separately because we need to create a new node

    public void insert(int key) {
        root = insert(root, new Node(key));
    }

    private Node insert(Node root, Node node) {
        if (root == null) return node;
        // check if node priority is > root priority,
        // if it is we call our split 
        // the point of calling split is to maintain the treap properties, we will set the left and right children of the new root to be the two treaps that we got from the split function, 
        if (node.priority > root.priority) {
            Node[] halves = split(root, node.key);

            node.left = halves[0];
            node.right = halves[1];

            return node;
        }
        // traverse the tree to maintain binary search tree order
        if (node.key < root.key) {
            root.left = insert(root.left, node);
        } else if (node.key > root.key) {
            root.right = insert(root.right, node);
        }
        return root;
    }

    /*
    the delete function is easy because all we do is merge
    the left and right subtrees of the node we want to delete,
    this merge will maintain the treap properties and we will return the new root of the merged treap,
    skipping over the node we want to delete, and then return all the way back up with the new treap
    */

    // again call it separately, same reason as insert
    public void delete(int key) {
        root = delete(root, key);
    }

    private Node delete(Node root, int key) {
        if (root == null) return null;

        // if found, remove node by merging its children
        if (root.key == key) {
            return merge(root.left, root.right);
        }

        // traverse to locate node, then refresh pointer
        if (key < root.key) {
            root.left = delete(root.left, key);
        } else {
            root.right = delete(root.right, key);
        }

        return root;
    }

    // search for key in treap (BST traversal)
    public boolean search(int key) {
        return search(root, key);
    }

    private boolean search(Node root, int key) {
        // normal search
        if (root == null) return false;
        if (root.key == key) return true;
        if (key < root.key) {
            return search(root.left, key);
        } else {
            return search(root.right, key);
        }
    }

    // inorder traversal for sorted output (by key)
    public void inorder() {
        inorder(root);
        System.out.println();
    }

    private void inorder(Node root) {
        if (root == null) return;
        inorder(root.left);
        System.out.print(root.key + " ");
        inorder(root.right);
    }

    // this function is just to help us test because 
    // our current inorder function only prints the values but we need a comparison
    private static String inorderString(treaps t) {
        return inorderString(t.root).trim();
    }

    private static String inorderString(Node root) {
        if (root == null) return "";
        String left = inorderString(root.left);
        String node = Integer.toString(root.key);
        String right = inorderString(root.right);

        String result = "";
        if (!left.isEmpty()) {
            result += left + " ";
        }
        result += node;
        if (!right.isEmpty()) result += " " + right;
        return result;
    }

    private static boolean runTest(String inputFile, String outputFile) {
        treaps treap = new treaps();
        List<String> actualOutput = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.trim().split("\\s+");
                if (parts.length == 0 || parts[0].isEmpty()) continue;

                String cmd = parts[0];
                if (cmd.equals("insert")) {
                    treap.insert(Integer.parseInt(parts[1]));
                } else if (cmd.equals("erase")) {
                    treap.delete(Integer.parseInt(parts[1]));
                } else if (cmd.equals("search")) {
                    boolean result = treap.search(Integer.parseInt(parts[1]));
                    actualOutput.add(Boolean.toString(result).toLowerCase());
                } else if (cmd.equals("inorder")) {
                    actualOutput.add(inorderString(treap));
                }
            }
        } catch (IOException e) {
            System.err.println("Failed to read input file: " + e.getMessage());
            return false;
        }

        List<String> expectedOutput = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(outputFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    expectedOutput.add(line.trim());
                }
            }
        } catch (IOException e) {
            System.err.println("Failed to read output file: " + e.getMessage());
            return false;
        }

        boolean passed = actualOutput.equals(expectedOutput);
        System.out.println(inputFile + ": " + (passed ? "PASS" : "FAIL"));
        if (!passed) {
            System.out.println("Expected:");
            expectedOutput.forEach(System.out::println);
            System.out.println("Got:");
            actualOutput.forEach(System.out::println);
        }

        return passed;
    }

    public static void main(String[] args) {
        Path baseDir = Paths.get(System.getProperty("user.dir"));
        Path ioDir = baseDir.resolve("io");

        int total = 3;
        int passed = 0;

        for (int i = 1; i <= total; i++) {
            String inFile = ioDir.resolve("sample.in." + i).toString();
            String outFile = ioDir.resolve("sample.out." + i).toString();
            if (runTest(inFile, outFile)) {
                passed++;
            }
        }

        System.out.println();
        System.out.println("Final: " + passed + " / " + total + " tests passed.");
        if (passed != total) {
            System.exit(1);
        }
    }
}
